include/exadg/operators/operator_base.cpp

```diff
@@ -1001,6 +1011,54 @@ OperatorBase<dim, Number, n_components>::internal_calculate_system_matrix(
 
   // communicate overlapping matrix parts
   system_matrix.compress(dealii::VectorOperation::add);
+
+  // set diagonal entries of constrained DoFs to 1.0
+  dealii::DoFHandler<dim> const & dof_handler =
+    this->matrix_free->get_dof_handler(this->data.dof_index);
+  for(auto const & line : this->constraint->get_lines())
+  {
+    if(dof_handler.locally_owned_dofs().is_element(line.index))
+    {
+      std::cout << "system_matrix.el(i,i) = " << system_matrix.el(line.index, line.index)
+                << std::endl;
+      system_matrix.set(line.index, line.index, 1.0);
+    }
+  }
+  system_matrix.compress(dealii::VectorOperation::insert);
+}
+
+template<int dim, typename Number, int n_components>
+void
+OperatorBase<dim, Number, n_components>::get_constant_modes(
+  std::vector<std::vector<bool>> &   constant_modes,
+  std::vector<std::vector<double>> & constant_modes_values) const
+{
+  (void)constant_modes_values;
+  std::cout << "OperatorBase::get_constant_modes ##+" << std::endl;
+  dealii::DoFHandler<dim> const & dof_handler =
+    this->matrix_free->get_dof_handler(this->data.dof_index);
+
+  std::cout << "this->level = " << this->level << std::endl;
+  std::cout << "this->matrix_free->get_mg_level() = " << this->matrix_free->get_mg_level()
+            << std::endl;
+  if(this->matrix_free->get_mg_level() != dealii::numbers::invalid_unsigned_int)
+  {
```

Comment:
if I use `get_mg_level()` to decide whether we are on the coarse grid or not, setting `bool adaptive_refinement = true` in the Poisson and Structure applications, I see different values irrespective of using GMG+AMG coarse solver or not.

I would say that using adaptive refinement should not impact the return value of `matrix_free.get_mg_level()` per se? But this is exactly what I see. Using adaptivity, I get
`matrix_free.get_mg_level() == dealii::numbers::invalid_unsigned_int`
, but for identical settings but not using adaptivity, I get
`matrix_free.get_mg_level() == 0` (coarse level AMG solver within GMG)
Any ideas what is going on @kronbichler ? Thanks!

Reply:
with adaptivity, I currently get with 4 cores:
![image](https://github.com/user-attachments/assets/5beac124-c86c-429d-a67a-2297428cace9)
1) the "WARNING" is shown `(n_mpi_ranks-1)`-times
2) I get `matrix_free.get_mg_level() == dealii::numbers::invalid_unsigned_int`, such that we enter the `else` branch that is intended for the fine level AMG solve.